### PR TITLE
sql/importer: wait for any pending ingestion to complete before rolling back

### PR DESCRIPTION
**Pull request by dt**
This change adds a new method the job registry, allowing any ingestion
activity related to a given job ID to be indicated to the job registry,
including ingestion activity that is being performed by a processor that
is being run on this node on behalf of a job being coordinated by some
other node.

The registry also exposes a method to query if a given job has indicated
to the registry on a given node that it is engaged in ingestion activity
on that node.

In conjunction, these APIs can allow a job to indicate when and where it
is running ingestion tasks and for inspection of this information by
other tasks, including potentially subsequent retries or phases of the
same job. Implementation of such usages is left to a future change.

This adds a new helper utility that can be used to check if a given job
has indicated that it is currently performing ingestion work on any node
using the job Registry's MarkAsIngesting API.

This is done by running a distSQL flow on all nodes that invokes a new
processor which simply reads this registry API, returning an error if
the given job ID is ingesting and no error if it is not.

This utility is intended for use when a job needs to positively verify
that ingestion activity for a given job -- typically itself in a previous
phase or retry -- has ceases before moving on to a new phase or retry.

Typically, a job only moves on to a new phase after the flow running its
ingestion activity returns after every ingesting processor exits. However
in some rare edge cases, it is possible for this not to be the case: a
very slow, delayed ingestion could still be flushing after the coordinator
for the job terminated and the job was adopted on a different node. In
such a case, that new execution of the job, potentially now in a reverting
phase, could race with this delayed ingestion. In the case of a revert,
such a race could lead to incomplete cleanup by the revert, as data may
be added by the delayed ingestion after the reverting phase processes a
given span.

Integration of this new utility into jobs that require it is left to a
future change.

Release note (bug fix): Fix a potential bug whereby a failed or
cancelled IMPORT could in some cases leave some of the imported rows
behind after it was cancelled, in the rare event that the writing
processes were slow enough to continue writing after the cleanup process
started. 

Epic: none.

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 4 of 6 files at r2, 1 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97071)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @dt, @msbutler, and @stevendanna)

___
*[`pkg/jobs/ingeststopped/ingesting_check_processor.go` line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/97071#-NOFj8Jx3x8o5CJS01zJ:-NOFj8Jx3x8o5CJS01zK:bkot9pz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8691d51a9b47427551f005e6f316399c51d87675/pkg/jobs/ingeststopped/ingesting_check_processor.go#L30)):*
> ```Go
> 			// This processor doesn't have any inputs to drain.
> 			InputsToDrain:        nil,
> 			TrailingMetaCallback: func() []execinfrapb.ProducerMetadata { return nil },
> ```

nit: `TrailingMetaCallback` can be left `nil`.
___
*[`pkg/jobs/ingeststopped/ingesting_check_processor.go` line 52 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/97071#-NOFitk73xb32_tzPwuF:-NOFitk8CwGAEYBNNLDa:b-ord3xc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8691d51a9b47427551f005e6f316399c51d87675/pkg/jobs/ingeststopped/ingesting_check_processor.go#L52)):*
> ```Go
> }
> 
> func (p *proc) ConsumerClosed() {
> ```

nit: you can omit this method and defer to the implementation by `ProcessorBase`.

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97071)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @msbutler, @stevendanna, and @yuzefovich)

___
*[`pkg/jobs/ingeststopped/ingesting_check_processor.go` line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/97071#-NOFj8Jx3x8o5CJS01zJ:-NOK_cgFCfVNIWJG_2WJ:b-fzgyxl) ([raw file](https://github.com/cockroachdb/cockroach/blob/8691d51a9b47427551f005e6f316399c51d87675/pkg/jobs/ingeststopped/ingesting_check_processor.go#L30)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `TrailingMetaCallback` can be left `nil`.
</blockquote></details>

Done. 
Thanks!
___
*[`pkg/jobs/ingeststopped/ingesting_check_processor.go` line 52 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/97071#-NOFitk73xb32_tzPwuF:-NOK_aV1B6NLLRvPEqzU:b-fzgyxl) ([raw file](https://github.com/cockroachdb/cockroach/blob/8691d51a9b47427551f005e6f316399c51d87675/pkg/jobs/ingeststopped/ingesting_check_processor.go#L52)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: you can omit this method and defer to the implementation by `ProcessorBase`.
</blockquote></details>

Done. 
Thanks!

File: pkg/jobs/registry.go [hunk 1]
```diff
@@ -1960,3 +1965,32 @@ func (r *Registry) TestingIsJobIdle(jobID jobspb.JobID) bool {
 	adoptedJob := r.mu.adoptedJobs[jobID]
 	return adoptedJob != nil && adoptedJob.isIdle
 }
+
+// MarkAsIngesting records a given jobID as actively ingesting data on the node
+// in which this Registry resides, either directly or via a processor running on
+// behalf of a job being run by a differnt registry. The returned function is to
+// be called when this node is no longer ingesting for that jobID, typically by
+// deferring it when calling this method. See IsIngestin ag.
```
Comment (msbutler) [inline pkg/jobs/registry.go:1973]: nit typo in comment.

**Review by msbutler — COMMENTED** [review]
this is awesome!

File: pkg/sql/rowexec/processors.go [hunk 2]
```diff
@@ -242,6 +242,15 @@ func NewProcessor(
 		}
 		return NewReadImportDataProcessor(ctx, flowCtx, processorID, *core.ReadImport, post, outputs[0])
 	}
+	if core.IngestStopped != nil {
+		if err := checkNumInOut(inputs, outputs, 0, 1); err != nil {
+			return nil, err
+		}
+		if NewIngestStoppedProcessor == nil {
+			return nil, errors.New("IsIngestingImport processor unimplemented")
```
[the remark this answers is not in the document]
Comment (stevendanna) [reply pkg/sql/rowexec/processors.go:250]: I believe the idea is that in mixed-version clusters, we just resign ourselves to waiting out the max wait time.

File: pkg/jobs/ingeststopped/ingesting_checker.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	ctx, sp := tracing.ChildSpan(ctx, "ingeststopped")
+	defer sp.Finish()
+
+	retries := retry.StartWithCtx(ctx, retry.Options{
+		InitialBackoff: time.Second,
+		MaxBackoff:     time.Second * 10,
+	})
+
+	started := timeutil.Now()
+	const maxIngestingCheckDelay = time.Minute * 5
```
Comment (stevendanna) [inline pkg/jobs/ingeststopped/ingesting_checker.go:48]: Is this unused?

File: pkg/jobs/ingeststopped/ingesting_checker.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	"github.com/cockroachdb/errors"
+)
+
+// WaitForNoIngestingNodes contacts all nodes in the cluster and verifies that
+// they are not ingesting data for the given job ID, and continues to try to do
+// so for the specified duration until it succeeds or times out.
+func WaitForNoIngestingNodes(
+	ctx context.Context, execCtx sql.JobExecContext, job *jobs.Job, maxWait time.Duration,
+) error {
+	ctx, sp := tracing.ChildSpan(ctx, "ingeststopped")
```
Comment (stevendanna) [inline pkg/jobs/ingeststopped/ingesting_checker.go:39]: I think we've been trying to use function names in these spans when it lines up one-to-one like this.

File: pkg/jobs/ingeststopped/ingesting_checker.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+		if timeutil.Since(started) > maxWait {
+			return err
+		}
+
+		if timeutil.Since(lastStatusUpdate) > statusUpdateFrequency {
+			if statusErr := job.NoTxn().RunningStatus(ctx, func(_ context.Context, _ jobspb.Details) (jobs.RunningStatus, error) {
+				return jobs.RunningStatus(fmt.Sprintf("waiting for all nodes to finish writing before reverting: %s", err)), nil
+			}); statusErr != nil {
+				log.Warningf(ctx, "failed to update running status of job %d: %s", errors.Safe(job.ID()), statusErr)
```
Comment (stevendanna) [inline pkg/jobs/ingeststopped/ingesting_checker.go:68]: [question] Did you find errors.Safe necessary here? I thought we implemented SafeValue on JobID's

[nit] We don't log the JobID in the Infof call above but do here. We could make them consistent.  In the only production caller of this code, I believe we'll have the job ID in the log tags.

**Review by stevendanna — COMMENTED** [review]
Overall this looks reasonable to me. Left a few nits but nothing major.

I'll leave it to you and @msbutler to decide whether proceeding-on-error is the right call for now. I think it probably is.

One note is that the jobs.Registry API talks about ingest, but I suppose we could name it differently to make it a bit more general. In case we want to use it elsewhere.

**Comment by dt** [comment]
I did a quick/messy backport of this on to 22.1 and then rebased #100588 on it and uncommented the QueryQueryResult in its test.

Without this diff: 
```
=== RUN   TestImportIntoCSVCancel
    test_log_scope.go:79: test logs captured to: /var/folders/7q/ky_c8cc958dfrv9w9sfql8sc0000gp/T/logTestImportIntoCSVCancel1785227974
    test_log_scope.go:80: use -show-logs to present logs inline
    import_stmt_test.go:2016: query 'SELECT count(*) FROM t': expected:
        0

        got:
        646

    import_stmt_test.go:2017: -- test log scope end --
98 runs completed, 1 failures, over 53s
```

With this diff:
`1793 runs so far, 0 failures, over 20m50s`

**Review by msbutler — APPROVED** [review]
Time for me to find another issue to bother people about!

**Comment by dt** [comment]
TFTR!

bors r+

**Pull request merged**